Allow advanced modifications of Amazon cloud volumes

gberginc opened the pull request:
Amazon EBS has recently introduced ability to modify cloud volume
configuration (type of cloud volume, its size and the IOPS). This patch
reuses the UI from the cloud volume creation dialog to allow users to
modify volumes on the fly.

Depends on
 * https://github.com/ManageIQ/manageiq-providers-amazon/pull/176
 * https://github.com/ManageIQ/manageiq-providers-amazon/pull/177

Video: http://x.k00.fr/7t19f
 * shows how different volume types are created
 * compares OpenStack and Amazon UIs
 * shows UI for editing standard (Magnetic), gp2 and io1 volumes
   * magnetic: only name is enabled, other info just shown
   * gp2: type and size can be changed; IOPS disabled
   * io1: all fields can be changed
 * video shows AWS console to present the results

@miq-bot add_label enhancement, storage

gberginc commented:
@AparnaKarve this is the form I talked about yesterday. Although it depends on the Amazon provider changes, I would appreciate any feedback.

AparnaKarve commented on app/assets/javascripts/controllers/cloud_volume/cloud_volume_form_controller.js:
```diff
@@ -14,7 +14,7 @@ ManageIQ.angular.app.controller('cloudVolumeFormController', ['$http', '$scope',
       { type: "st1", name: "Throughput Optimized HDD (ST1)" },
       { type: "sc1", name: "Cold HDD (SC1)" },
-      { type: "magnetic", name: "Magnetic" },
+      { type: "standard", name: "Magnetic" },
```
Would have really liked if we could fetch this list on-the-fly using some API (AWS API perhaps?)
But don't worry about it, if that's not doable.

gberginc replied:
AWS does not provide an API for this; we initially thought about making `CloudVolumeType` model where storage managers would be able to list their options (for example in OpenStack admins are allowed to define different volume types so it would make sense).

My concern is that there are specific rules in the UI that need to be handled with regards to the chosen volume type, so I am not sure if providing it through API/backend would benefit much. Still, we could use something like this for [regions](https://github.com/ManageIQ/manageiq-providers-amazon/blob/master/app/models/manageiq/providers/amazon/regions.rb) or [instance types](https://github.com/ManageIQ/manageiq-providers-amazon/blob/master/app/models/manageiq/providers/amazon/instance_types.rb). Would that make any sense?

AparnaKarve replied:
Sure, yes. I know you mentioned this even before.

AparnaKarve commented on app/views/cloud_volume/_common_new_edit.haml:
```diff
@@ -0,0 +1,60 @@
+.form-group{"ng-class" => "{'has-error': angularForm.name.$invalid}"}
```
Can you rename this haml to `_form.html.haml`?

gberginc replied:
Yes, I missed that completely.

AparnaKarve commented on app/views/cloud_volume/edit.html.haml:
```diff
@@ -1,18 +1,9 @@
+    = render :partial => "common_new_edit"
 
   %table{:width => '100%'}
     %tr
```
In this haml, down below, can you use this partial to render the buttons?
`"layouts/angular/x_edit_buttons_angular"`
instead of rendering the buttons individually.
(I probably missed telling you this in your earlier PR)

gberginc replied:
Yes, sure - I didn't know about this before.

gberginc replied:
@AparnaKarve I noticed there are some differences between the way how the button table is created for "new volume" than in the about partial layout. For example, in cloud volume case it checks if `@volume.id.nil?` to decide whether to show `Add` or `Save`, while the layout partial uses `newRecord`.

I guess we should update the whole controller/view to make it compliant with the partial layout. Perhaps in a separate PR. What do you think?

AparnaKarve commented on app/views/cloud_volume/new.html.haml:
```diff
@@ -47,64 +47,7 @@
-        %span.help-block{"ng-show" => "cloudVolumeModel.aws_volume_type == 'io1' && angularForm.aws_iops.$error.required"}
-          = _("Required")
+    = render :partial => "common_new_edit"
 
```
Same comment as above for the button partial

AparnaKarve commented:
@gberginc I have tested this in the UI and it works great!
Overall, this looks good other than some of the few minor changes mentioned above. Thanks.

AparnaKarve commented on app/assets/javascripts/controllers/cloud_volume/cloud_volume_form_controller.js:
```diff
@@ -33,7 +33,12 @@ ManageIQ.angular.app.controller('cloudVolumeFormController', ['$http', '$scope',
+    $scope.cloudVolumeModel.size = data.size;
+    // Currently, this is only relevant for AWS volumes so we are prefixing the
+    // model attribute with AWS.
+    $scope.cloudVolumeModel.aws_volume_type = data.volume_type;
```
@gberginc We now have REST API support for the `cloud_volumes` collection...which means we can get rid of this `$http.get` call here and replace it with --
```url
/api/cloud_volumes/<id>?attributes=ext_management_system.type
```

gberginc commented:
@AparnaKarve I've renamed the haml file, replaced `$http` with `API` and removed the corresponding call in the controller (+ route) since it's not used elsewhere.

I am not sure about the buttons at the bottom if it makes sense to change it within this PR, so I'd appreciate your feedback.

AparnaKarve commented:
> replaced $http with API and removed the corresponding call in the controller (+ route) since it's not used elsewhere.

Awesome! I reviewed that part and it looks good.

> I am not sure about the buttons at the bottom if it makes sense to change it within this PR

A separate follow-up PR later should be fine. Not a problem.

gberginc commented:
@AparnaKarve is there anything more to change in this one?

AparnaKarve commented:
@gberginc I think everything is addressed here other than the buttons, but there will be a follow-up PR for that.

@h-kataria Can you please merge this?
